src/internal/pfsdb/repos.go

```diff
@@ -0,0 +1,363 @@
+)
+
+const (
+	getRepoAndBranches = "SELECT repo.id, repo.name, repo.type, repo.project_id, " +
+		"repo.description, array_agg(branch.proto) AS branches, repo.created_at, repo.updated_at, project.name AS proj_name FROM pfs.repos repo " +
+		"JOIN core.projects project ON repo.project_id = project.id " +
+		"LEFT JOIN collections.branches branch ON project.name || '/' || repo.name || '.' || repo.type = branch.idx_repo "
+	noBranches = "{NULL}"
+)
+
+// ErrRepoNotFound is returned by GetRepo() when a repo is not found in postgres.
+type ErrRepoNotFound struct {
+	Project string
+	Name    string
+	Type    string
+	ID      pachsql.ID
+}
+
+// Error satisfies the error interface.
+func (err ErrRepoNotFound) Error() string {
+	if id := err.ID; id != 0 {
+		return fmt.Sprintf("repo id=%d not found", id)
+	}
+	return fmt.Sprintf("repo %s/%s.%s not found", err.Project, err.Name, err.Type)
+}
+
+func (err ErrRepoNotFound) GRPCStatus() *status.Status {
+	return status.New(codes.NotFound, err.Error())
+}
+
+func IsErrRepoNotFound(err error) bool {
+	return errors.As(err, &ErrRepoNotFound{})
+}
+
+// ErrRepoAlreadyExists is returned by CreateRepo() when a repo with the same name already exists in postgres.
+type ErrRepoAlreadyExists struct {
+	Project string
+	Name    string
+	Type    string
+}
+
+// Error satisfies the error interface.
+func (err ErrRepoAlreadyExists) Error() string {
+	if n, t := err.Name, err.Type; n != "" && t != "" {
+		return fmt.Sprintf("repo %s.%s already exists", n, t)
+	}
+	return "repo already exists"
+}
+
+func (err ErrRepoAlreadyExists) GRPCStatus() *status.Status {
+	return status.New(codes.AlreadyExists, err.Error())
+}
+
+func IsErrRepoAlreadyExists(err error) bool {
+	targetErr := &pgconn.PgError{}
+	ok := errors.As(err, targetErr)
+	if !ok {
+		return false
+	}
+	return targetErr.Code == "23505" // duplicate key SQLSTATE
+}
+
+// RepoIterator batches a page of repoRow entries. Entries can be retrieved using iter.Next().
+type RepoIterator struct {
+	limit    int
+	offset   int
+	repos    []repoRow
+	index    int
+	tx       *pachsql.Tx
+	where    string
+	whereVal interface{}
+}
+
+type repoRow struct {
+	ID          pachsql.ID `db:"id"`
+	Name        string     `db:"name"`
+	ProjectID   string     `db:"project_id"`
+	ProjectName string     `db:"proj_name"`
+	Description string     `db:"description"`
+	RepoType    string     `db:"type"`
+	CreatedAt   time.Time  `db:"created_at"`
+	UpdatedAt   time.Time  `db:"updated_at"`
+	// Branches is a string that contains an array of hex-encoded branchInfos. The array is enclosed with curly braces.
+	// Each entry is prefixed with '//x' and entries are delimited by a ','
+	Branches string `db:"branches"`
+}
+
+// NextWithID is like Next, but populates the 'id' and 'dst' parameters each iteration where 'id' is a pointer to the row ID and dst is the repoInfo.
+func (iter *RepoIterator) NextWithID(ctx context.Context, id *pachsql.ID, dst **pfs.RepoInfo) error {
```

Comment:
I think you should define a type something like

```go
type RepoPair struct {
  ID RepoID
  Repo RepoInfo
}
```

And change Next to return one of these. `Next(ctx context.Context, dst *RepoPair) error`

It's easier to get rid of information than add it.  And `NextWithID` can't be used with any of the utilities in `stream`.  The more useful method should be `Next`.